src/hotspot/share/classfile/protectionDomainCache.cpp

```diff
@@ -128,108 +159,69 @@ void ProtectionDomainCacheTable::unlink() {
 
   // Purge any deleted entries outside of the SystemDictionary_lock.
   purge_deleted_entries();
+  int oops_removed = purge_entries_from_table();
 
-  MutexLocker ml(SystemDictionary_lock);
-  int oops_removed = 0;
-  for (int i = 0; i < table_size(); ++i) {
-    ProtectionDomainCacheEntry** p = bucket_addr(i);
-    ProtectionDomainCacheEntry* entry = bucket(i);
-    while (entry != NULL) {
-      oop pd = entry->object_no_keepalive();
-      if (pd != NULL) {
-        p = entry->next_addr();
-      } else {
-        oops_removed++;
-        LogTarget(Debug, protectiondomain, table) lt;
-        if (lt.is_enabled()) {
-          LogStream ls(lt);
-          ls.print_cr("protection domain unlinked at %d", i);
-        }
-        entry->literal().release(Universe::vm_weak());
-        *p = entry->next();
-        free_entry(entry);
-      }
-      entry = *p;
-    }
-  }
   _total_oops_removed += oops_removed;
   _dead_entries = false;
 }
 
-void ProtectionDomainCacheTable::print_on(outputStream* st) const {
+void ProtectionDomainCacheTable::print_on(outputStream* st) {
   assert_locked_or_safepoint(SystemDictionary_lock);
-  st->print_cr("Protection domain cache table (table_size=%d, classes=%d)",
-               table_size(), number_of_entries());
-  for (int index = 0; index < table_size(); index++) {
-    for (ProtectionDomainCacheEntry* probe = bucket(index);
-                                     probe != NULL;
-                                     probe = probe->next()) {
-      st->print_cr("%4d: protection_domain: " PTR_FORMAT, index, p2i(probe->object_no_keepalive()));
-    }
-  }
+  auto printer = [&] (WeakHandle& key, WeakHandle& value) {
+      st->print_cr("  protection_domain: " PTR_FORMAT, p2i(value.peek()));
+  };
+  st->print_cr("Protection domain cache table (table_size=%d, protection domains=%d)",
+                _pd_cache_table.table_size(), _pd_cache_table.number_of_entries());
+  _pd_cache_table.iterate_all(printer);
 }
 
 void ProtectionDomainCacheTable::verify() {
-  verify_table<ProtectionDomainCacheEntry>("Protection Domain Table");
-}
-
-oop ProtectionDomainCacheEntry::object() {
-  return literal().resolve();
+  auto verifier = [&] (WeakHandle& key, WeakHandle& value) {
+    guarantee(value.peek() == nullptr || oopDesc::is_oop(value.peek()), "must be an oop");
+  };
+  _pd_cache_table.iterate_all(verifier);
 }
 
 // The object_no_keepalive() call peeks at the phantomly reachable oop without
 // keeping it alive. This is okay to do in the VM thread state if it is not
```

Comment:
You don't call `object_no_keepalive()` any more

Reply:
This one (not the one I removed), is called by dictionary.cpp - the pd_set is a linked list of ProtectionDomainEntry, where we don't keep the WeakHandle alive when looking at the value.

Reply:
I don't understand your point. My point is that the comment no longer seems correct.

Reply:
The first sentence of the comment is correct.  I don't know what the second sentence means, so I'll remove it.